### DISCUSSION TITLE
Universal middleware

### DISCUSSION THREAD

**Discussion by cyco130** [hattipjs/hattip#45] [Ideas]
Previous discussion was here: https://github.com/honojs/hono/issues/443

**Comment by mindplay-dk** [comment]
So just to reboot the conversation: I'd like to devise something that relies on dependency injection (a proper platform abstraction) which would simplify both testing and bootstrapping in general.

So here's my current thinking:

```ts
// Platform features (file IO etc) is made available to Handlers:

type Platform = {
  glob: (pattern: string) => FileInfo[];
  open: (path: string) => ReadableStream;
  // ...
}

interface FileInfo {
  path: string;
  // ...
}

// Context provides Request-specific dependencies, communication between handlers:

type Context = Record<symbol, unknown>;

// Handlers bootstrap with the Platform, then process Requests and produce Responses:

type Handler = (platform: Platform) => (request: Request, context: Request) => Response;
```

Thoughts?

**Comment by cyco130** [comment]
I think a full platform standardization might be out of reach today. But there *is* a [proposed web standard](https://developer.mozilla.org/en-US/docs/Web/API/File_and_Directory_Entries_API) for a filesystem API. I'm not very familiar with it but it might be a starting point.

**Reply by mindplay-dk** [reply]
No, I don't envision building a complete platform standard - just enough to support interesting and useful middleware, such as a static file server, a disk-based cache or proxy, and so on.

You don't need a ton of things to support those, I don't think?

* `glob(path, pattern): string[]` to enumerate files in a logical, local file-system
* `openRead(path): ReadableStream`
* `openWrite(path): WritableStream`
* `fetch` for basic network IO
* something for de/compression of streams

We're already basing these new frameworks on the `Request` and `Response` models from `fetch` anyway, and the [streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) lives in the same neck of the woods.

Adding support for the `FileSystem` API might not be totally out of reach though - there's a [ponyfill](https://github.com/use-strict/file-system-access) with support for Node and Deno.

I mean, I suppose you could argue the individual handler or middleware could just ship with a factory function, and inject the functions they need - so for example, a static file server might need to be bootstrapped with `glob` and `open` functions, and a set of implementations for whatever platforms it currently supports.

But do you think there's enough value in having a standard for handlers then? Most complex middleware would be platform dependent anyway. If effectively all we're standardizing is `(Request) => Awaitable<Response>` then, as you said previously, that's already a de-facto standard for some projects. If a static file server has to ship with it's own platform-dependent abstractions anyway, what value does this standard provide?

Just a set of framework adapters for `(Request) => Awaitable<Response>`?

That might have some merit, but it doesn't really seem to be where most of the work (problems) lie for most complex middleware?

**Reply by cyco130** [reply]
OK, I think a set of minimal filesystem API adapters might be a good starting point before thinking about how to inject the API. It would be a generally useful utility not tied to HatTip or Hono anyway. Obvious candidates for starters:

- Memory (useful for testing but maybe also for serving small, frequently requested files like favicons)
- Node `fs`
- Deno `std/fs`
- Cloudflare Workers KV

And for later:

- S3 (or compatible, like Cloudflare R2)
- Redis and similar KV stores
- Browser `FileSystem` and/or Indexed DB (maybe useful for service workers)

**Comment by masx200** [comment]
Can we design an adapter from Koa middleware to Hattip middleware?

**Comment by magne4000** [comment]
I'm currently really interested in solving this issue about universal middlewares (or at least request handlers).

I'm working on [Bati](https://batijs.dev/) which is [Vike](https://vike.dev/)'s new pick-and-choose scaffolder. In Bati, you can choose between multiple server libs, and currently each middleware need to be written manually for each server as there is no common way to represent them.
As you can guess, it's hard to scale and to maintain such architecture, so I'm willing to tackle this issue.

@mindplay-dk I liked the idea of [universal handlers](https://github.com/honojs/hono/issues/443#issuecomment-1445042681), and I think that solving this issue first reduces the scope of maintainability quite drastically in Bati's case.

With universal handlers out of the way, writing custom middlewares for each frameworks could be solved by a lib that creates the right middleware for you (like [unplugin](https://github.com/unjs/unplugin) does for builders).

I'll soon try to create an "any server middleware" to "generic handler + context" (and vice-versa) as some kind of POC within ~~Bati's codebase~~ a dedicated repo and link it here so that we can discuss real practical implementations, be aware and list all limitations that we can think of, and try to solve them one by one.

NB: Here are the current servers implementations in Bati:
- [Hono](https://github.com/batijs/bati/blob/5c404396e9841e6fdae60c19956d8a83da6b7e7a/boilerplates/hono/files/hono-entry.ts)
- [HatTip](https://github.com/batijs/bati/blob/75df6894a2c42ba6cbeb953c0bb5d342f91caac0/boilerplates/hattip/files/hattip-entry.ts)
- [h3](https://github.com/batijs/bati/blob/a04c72ec1d80502a2581cff510699fce27612973/boilerplates/h3/files/h3-entry.ts)
- [express](https://github.com/batijs/bati/blob/a04c72ec1d80502a2581cff510699fce27612973/boilerplates/express/files/express-entry.ts)

**Reply by magne4000** [reply]
cc @phonzammi who helped writing some server implementations and some middlewares

**Reply by brillout** [reply]
I think this has the potential to become foundational.

Eventually, tools could include a minimal/zero-config universal server integration. E.g. `vike/unserver`, `telefunc/unserver`, `@next-auth/unserver`, `rakkasjs/unserver`, etc.

How about the name? Ideas? The npm package `unserver` and `unserve` are available, the org `unserver` unfortunately isn't. The org `@unserve` was available and I went ahead and picked it. I'm thinking an npm org would be neat, e.g. for `@unserve/next-auth` until there is an official `@next-auth/unserve`. Or we just use an npm prefix `unserver-next-auth`; I expect the prefix to be always-ish available.

**Reply by brillout** [reply]
(This has a lot in common with the Stem vision. Maybe it should be part of Stem, or maybe it should be a separate project. I'll think more about it once I'm done with my current priorities. In the meantime having a POC would be quite helpful.)

**Reply by magne4000** [reply]
> Eventually, tools could include a minimal/zero-config universal server integration. E.g. vike/unserver, telefunc/unserver, @next-auth/unserver, rakkasjs/unserver, etc.

That's what I was thinking too, could be really cool if we manage that.

> How about the name?

I was also thinking about `unserver`, but `unserve` is good to.

**Reply by brillout** [reply]
Actually, how about one of the two?
 - `unimid` / **Uni**versal **Mid**llware
 - `unhandler` / **Un**iversal **Handler**

The npm packages and orgs are available.

While the name "unserver" is appealing, it's maybe a bit confusing as the project doesn't aim to create a universal server but universal middlewares/handlers. E.g. it's "unplugin" not "unbundler".

WDYT?

**Reply by magne4000** [reply]
The most widespread term right now in the whole ecosystem is Middleware IMO, even when we talk about Handlers, so I would be more in favor of `unimid`, and I like the sound of it.

> E.g. it's "unplugin" not "unbundler".

True

**Reply by brillout** [reply]
:+1: 

> The most widespread term right now in the whole ecosystem is Middleware

Same thought and unfortunately `unmiddleware`/`unimiddleware` is a bit (too?) long. 

The thing I like about `unhandler` is that it's clear yet not too long, and there is something appealing to it. Or maybe even `unihandler`, that's even clearer but I find it a bit less appealing. I also like `unimid` although it isn't clear on its own.

Anyways, all these options are available as npm pkg & org, so I guess we can go with any of these options.

**Reply by brillout** [reply]
> I would be more in favor of `unimid`, and I like the sound of it.

I also find that pronoucing `unimid` out loud sounds nice.

> `unhandler` [...] there is something appealing to it.

More precisely: visually appealing.

**Reply by brillout** [reply]
Actually, maybe `universal-middleware` could be an option.

```js
// server.js

import vike from 'vike/universal-middleware'
import telefunc from 'vike/universal-middleware'
import universalMiddleware from 'universal-middleware/express'

app.use(universalMiddleware([vike, telefunc]))
```

Annoying to type, but crystal clear.

Versus:

```js
// server.js

import vike from 'vike/unimid'
import telefunc from 'vike/unimid'
import unimid from 'unimid/express'

app.use(unimid([vike, telefunc]))
```

I'd say let's go with what you feel like.

**Reply by mindplay-dk** [reply]
@brillout I like it! there's enough obscure branding in the JS ecosystem already.

`universal-middleware` is as clear and unfancy as it can be. 😄

**Reply by mindplay-dk** [reply]
`universal-handler` (what @magne4000 is going with) is equally fine btw 🙂👍

**Comment by magne4000** [comment]
I created a repository to compile some ideas and some code. Feel free to open discussions are create PRs for specific topics.
https://github.com/magne4000/universal-handler
